buffer.go

```diff
@@ -0,0 +1,200 @@
+		c.BroadcastRenameEvent(oldNameEvent)
+		return
+	} else if newNameExist && !oldNameExist { //ITEM CREATED
+		newNameEvent.Flags = ItemCreated
+	} else if newNameExist && oldNameExist { //RENAME WITH FULL INFO
+		newNameEvent.OldPath = oldNameEvent.Path
+	}
+	c.BroadcastRenameEvent(newNameEvent)
+}
+func (c *Cache) CheckForMatch(eventId uint64) (event Event, matchedEvent Event, eventExist, matchExist bool, mode string) {
+	if event, eventExist = c.RenameFrom[eventId]; eventExist {
+		delete(c.RenameFrom, event.ID)
+		matchedEvent, matchExist = c.RenameTo[event.ID+1]
+		if matchExist {
+			delete(c.RenameTo, matchedEvent.ID)
+		}
+		mode = "RENAME_FROM"
+		return
+	} else if event, eventExist = c.RenameTo[eventId]; eventExist {
+		delete(c.RenameTo, event.ID)
+		matchedEvent, matchExist = c.RenameFrom[event.ID-1]
+		if matchExist {
+			delete(c.RenameTo, matchedEvent.ID)
+		}
+		mode = "RENAME_TO"
+		return
+	}
+	return
+}
+func (c *Cache) removeHead() {
+	event, matchedEvent, eventExist, matchExist, mode := c.CheckForMatch(c.timers.Head().ID)
+	if mode == "RENAME_TO" {
+		c.createRenameEvent(matchedEvent, event, matchExist, eventExist)
+		c.timers.moveHead()
+	} else if mode == "RENAME_FROM" {
+		c.createRenameEvent(event, matchedEvent, eventExist, matchExist)
+		c.timers.moveHead()
+	}
+}
+func (c *Cache) addToTimer(eventId uint64) {
+	for c.timers.NumberOfItems >= c.timers.Capacity {
+		c.removeHead()
+	}
+	c.timers.Add(time.Now(), eventId)
+	c.timers.Next()
+}
+func (c *Cache) timeDifference(t time.Time) int {
+	currentTime := time.Now()
+	difference := currentTime.Sub(t)
+	total := int(difference.Milliseconds())
+	return total
+}
+func (c *Cache) EventExists(eventId uint64) bool {
+	//TODO maybe return the event
+	if _, exist := c.RenameFrom[eventId]; exist {
+		return true
+	} else if _, exist := c.RenameTo[eventId]; exist {
+		return true
+	}
+	return false
+}
+func (c *Cache) Add(e Event, mode string) {
+	eventId := e.ID
+	if mode == "RENAME_TO" {
+		if renameFromEvent, exist := c.RenameFrom[eventId-1]; exist {
+			c.createRenameEvent(renameFromEvent, e, exist, true)
+			delete(c.RenameFrom, renameFromEvent.ID)
+		} else {
+			c.RenameTo[e.ID] = e
+			c.addToTimer(e.ID)
+		}
+	} else {
+		if renameToEvent, exist := c.RenameTo[eventId+1]; exist {
+			c.createRenameEvent(e, renameToEvent, true, exist)
+			delete(c.RenameTo, renameToEvent.ID)
+		} else {
+			c.RenameFrom[e.ID] = e
+			c.addToTimer(e.ID)
+		}
+	}
+}
+func (c *Cache) getEvent(eventId uint64, mode string) (reqEvent Event, exist bool) {
+
+	if mode == "RENAME_TO" {
+		if reqEvent, exist = c.RenameTo[eventId]; exist {
+			delete(c.RenameTo, reqEvent.ID)
+		}
+		return
+	} else {
+		if reqEvent, exist = c.RenameFrom[eventId]; exist {
+			delete(c.RenameFrom, reqEvent.ID)
+		}
+		return
+	}
+}
+func (c *Cache) findExpiredEvents() {
+	itemRemoved := true
+	for c.timers.NumberOfItems > 0 && itemRemoved {
+		if c.timeDifference(c.timers.Head().T)*1000 > int(1*time.Millisecond) {
```

Comment:
One cool trick is just to convert `c.timeDifference` to `int`, then you can `*1000` and `> 1000` instead of playing with `time.Duration`.